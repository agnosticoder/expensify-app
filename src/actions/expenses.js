import uuid from 'uuid';
import database from '../firebase/firebase';


//================================ADD_EXPENSE
const addExpense = (expense) => {
    return{
        type: 'ADD_EXPENSE',
        expense
    }
};

//To add the redux-thunk functionality - this return the function as oppose to object
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;

        const expense = { description, note, amount, createdAt};

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        })
    }
};

//==============================REMOVE_EXPENSE
const removeExpense = ({id = ''} = {}) => {
    return{
        type: 'REMOVE_EXPENSE',
        expense: {
            id: id
        } 
    }
};
//==============================EDIT_EXPENSE
const editExpense = (id, updates) => {
    return{
        type: 'EDIT_EXPENSE',
        id: id,
        updates: updates
    }
};

export { addExpense, removeExpense, editExpense};

//SET_EXPENSES
export const setExpenses = (expenses) => {
    return {
        type: 'SET_EXPENSES',
        expenses
    }
};

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((expensesFirebase) => {
            const expensesStore = [];
            expensesFirebase.forEach((expense) => {
                expensesStore.push({
                    id: expense.key,
                    ...expense.val()
                });
            });
            dispatch(setExpenses(expensesStore));   
        });
    };
};