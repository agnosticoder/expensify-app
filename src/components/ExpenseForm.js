import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker} from 'react-dates';

class ExpenseForm extends React.Component {
    constructor(props){
        super(props);

        this.onDescriptionChange = this.onDescriptionChange.bind(this);
        this.onNoteChange = this.onNoteChange.bind(this);
        this.onAmountChange = this.onAmountChange.bind(this);
        this.onDateChange = this.onDateChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);

        this.state = {
            description: props.expense ?  props.expense.description : '',
            note: props.expense ?  props.expense.note : '',
            amount: props.expense ?  ((props.expense.amount)/100).toString() : '',
            createdAt: props.expense ?  moment(props.expense.createdAt) : moment(),
            focused: false,
            error: ''
        };

    }

    onDescriptionChange(e){
        const description = e.target.value;
        this.setState(() => {
            return{
                description: description
            }   
        });
    }

    onNoteChange(e){
        const note = e.target.value;
        this.setState(()=>{
            return{
                note: note
            }
        });
    }

    onAmountChange(e){
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => {
                return{
                    amount: amount
                }
            });
        };
    }

    onDateChange(date){
        if(date){
            this.setState({ createdAt: date })
        }
    }

    onSubmit(e){
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            //Set error state equal to 'Please provide description and amount'
            this.setState(()=> {
                return{
                    error: 'Please provide description and amount'
                }
            })
        }else{
            this.setState(()=> {
                return{
                    error: ''
                }
            })
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount, 10)*100, // in cents not dollars
                createdAt: this.state.createdAt.valueOf() 
            });
        }
    }

    render() {
        return (
            <div>
                {!!this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type='text' placeholder='Description' autoFocus value={this.state.description}onChange={this.onDescriptionChange} />
                    <input type='text' placeholder='Amount' value={this.state.amount} onChange={this.onAmountChange} autoFocus />
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={this.onDateChange} // PropTypes.func.isRequired
                        focused={this.state.focused} // PropTypes.bool
                        onFocusChange={({ focused }) => this.setState(() => ({ focused }))} // PropTypes.func.isRequired
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea placeholder='Add a note for your expense(optional)' value={this.state.note} onChange={this.onNoteChange}></textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}



export default ExpenseForm;