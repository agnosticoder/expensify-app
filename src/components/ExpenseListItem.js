import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = (props) => {
    return (
        <div key={props.expense.id}>
            <Link to={`/edit/${props.expense.id}`}>
            <h3>Description: {props.expense.description}</h3>
            </Link>
            <p>Amount: {props.expense.amount} - Created At: {props.expense.createdAt}</p>
            {/* <li>Notes: {props.expense.note}</li> */}
        </div>
    );
};

export default ExpenseListItem;