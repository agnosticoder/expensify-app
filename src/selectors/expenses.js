import moment from 'moment';
// Timestamps (milliseconds)
// 0 timestamp = 1st January, 1970, 12AM (unix epoch)
//==============================================================================Get Filtered Expenses
const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1;
        }if(sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1;
        }
    });   
};

export default getVisibleExpenses;