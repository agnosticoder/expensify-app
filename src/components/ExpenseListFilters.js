import React from 'react';
import { connect } from 'react-redux';
import 'react-dates/initialize';
import { DateRangePicker} from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';


export class ExpenseListFilters extends React.Component{
    constructor(props){
        super(props);

        this.onDatesChange = this.onDatesChange.bind(this);

        this.state = {
            focusedInput: null
        }
    }

    onDatesChange({startDate, endDate}){
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    

    render() {
        return (
            <div>
                <input type='text' value={this.props.filters.text} onChange={(e) => {
                    this.props.setTextFilter(e.target.value);
                }} />
                <select value={this.props.filters.sortBy} onChange={(e) => {
                    if (e.target.value === 'date') {
                        this.props.sortByDate();
                    } else {
                        this.props.sortByAmount();
                    }
                }}>
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                    startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                    endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                    endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={focusedInput => this.setState({ focusedInput })} // PropTypes.func.isRequired,
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates= {true}
                />
            </div>
        );

    }
}

const getState = (state) => {
    return  {
        filters: state.filters
    };
};

const mapDispatchToProp = (dispatch) => {
    return {
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate)),
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount())
    }
};

export default connect(getState, mapDispatchToProp)(ExpenseListFilters);