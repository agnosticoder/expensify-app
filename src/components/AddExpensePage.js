import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export const AddExpensePage  = (props) => {
    return (
        <div>
            <h1>Expense Form</h1>
            <ExpenseForm onSubmit={(expense) => {
                //props.dispatch(addExpense(expense));
                props.addExpense(expense);
                props.history.push('/');
            }} />
        </div>
    );
};

const mapDispatchToProp = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    }
};

export default connect(undefined, mapDispatchToProp)(AddExpensePage);