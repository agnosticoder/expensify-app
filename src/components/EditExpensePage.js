import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';



export const EditExpensePage = (props) => {
    return (
        <div>
            Wana do some change with id {props.match.params.id}
            <ExpenseForm expense={props.expense}
                onSubmit={(expense) => {
                    props.editExpense(props.expense.id, expense);
                    props.history.push('/');
                }}
            />
            <button onClick={()=> {
                props.removeExpense({id: props.expense.id});
                props.history.push('/');
            }}>
            Remove
            </button>
         </div>
    );
};

const getState = (state,props) => {
    return{
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    }
};

const matchDispatchToProp = (dispatch, ownPorps) => {
    return {
        editExpense: (id,expense) => dispatch(editExpense(id, expense)),
        removeExpense: (data) => dispatch(removeExpense(data))
    }
};

export default connect(getState, matchDispatchToProp)(EditExpensePage);