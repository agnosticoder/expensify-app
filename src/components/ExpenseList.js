import React from 'react';
import { connect } from 'react-redux'
import getVisibleExpenses from '../selectors/expenses';
import ExpenseListItem from './ExpenseListItem';

export const ExpenseList = (props) => {
    return (
        <div>
            {props.expenses.length === 0 ? (
            <p>No Expenses</p>
            ):(
            props.expenses.map(expense => {
                return (<ExpenseListItem key={expense.id} expense={expense} />);
            })
            )}
        </div>
    );
};


//=======================================================================================================
//that is the new part that is resposible for passing the state as a prop ans dispatch function as a prop
//we can use these state and dispatch function to render the stuff or change the stuff we want
//We dont need to pass the state prop from the class component state, we are directly accessing it from 
// redux store -- cool huh!
//=======================================================================================================  

//What we gonna need from the store, state arg containes the entire redux store state
const getState = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}

//Connect return function - what and pass out comp to that fn as an argument
export default connect(getState)(ExpenseList);