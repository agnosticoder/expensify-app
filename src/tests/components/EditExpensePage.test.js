import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';
import { wrap } from 'module';

test('should render EditExpensePage correctly', () => {
    const matchSpy = { params : {
        id: 'This is some id'
    }}
    const wrapper = shallow(<EditExpensePage match={matchSpy} />);
    expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {
    const matchSpy = { params : {
        id: 'This is some id'
    }};
    const expenseSpy = {
        id: 'oh god'
    };

    const editExpenseSpy = jest.fn();

    const historySpy = { push: jest.fn()};

    const wrapper = shallow(
    <EditExpensePage 
    match={matchSpy} 
    expense={expenseSpy} 
    editExpense={editExpenseSpy} 
    history={historySpy} />);

    wrapper.find("ExpenseForm").prop('onSubmit')(expenses[1]);
    expect(editExpenseSpy).toHaveBeenCalledWith('oh god', expenses[1]);
    expect(historySpy.push).toHaveBeenCalledWith('/');
});


test('should handle removeExpense', () => {
    const matchSpy = { params : {
        id: 'This is some id'
    }};
    const expenseSpy = {
        id: 'oh god'
    };

    const removeExpenseSpy = jest.fn();

    const historySpy = { push: jest.fn()};

    const wrapper = shallow(
    <EditExpensePage 
    match={matchSpy} 
    expense={expenseSpy} 
    removeExpense={removeExpenseSpy}
    history={historySpy} />);

    wrapper.find('button').simulate('click');
    expect(removeExpenseSpy).toHaveBeenCalledWith({id: 'oh god'});
    expect(historySpy.push).toHaveBeenCalledWith('/');
});