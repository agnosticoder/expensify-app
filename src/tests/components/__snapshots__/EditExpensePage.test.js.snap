// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  Wana do some change with id 
  This is some id
  <ExpenseForm
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
